src/ophys_etl/utils/video_utils.py

```diff
@@ -91,3 +92,289 @@ def encode_video(video: np.ndarray, output_path: str,
+            read_data = in_file['data'][()]
+            min_max = np.quantile(read_data, quantiles)
+            data = read_data[:,
+                             origin[0]:origin[0]+frame_shape[0],
+                             origin[1]:origin[1]+frame_shape[1]]
+            del read_data
+        else:
+            data = in_file['data'][:,
+                                   origin[0]:origin[0]+frame_shape[0],
+                                   origin[1]:origin[1]+frame_shape[1]]
+
+    if min_max[0] > min_max[1]:
+        raise RuntimeError(f"min_max {min_max} in "
+                           "_read_and_scale_all_at_once; "
+                           "order seems to be reversed")
+
+    data = normalize_array(data,
+                           lower_cutoff=min_max[0],
+                           upper_cutoff=min_max[1])
+    return data
+
+
+def _read_and_scale_by_chunks(
+        full_video_path: Path,
+        origin: Tuple[int, int],
+        frame_shape: Tuple[int, int],
+        quantiles: Optional[Tuple[int, int]] = None,
+        min_max: Optional[Tuple[int, int]] = None,
+        time_chunk_size: int = 100) -> np.ndarray:
+
+    """
+    Read in a video from an HDF5 file and scale it to np.uint8
+    one chunk at a time
+
+    Parameters
+    ----------
+    full_video_path: pathlib.Path
+        Path to the HDF5 file
+
+    origin: Tuple[int, int]
+        Origin of the desired field of view
+
+    frame_shape: Tuple[int, int]
+        Shape of the desired field of view
+
+    quantiles: Optional[Tuple[float, float]]
+        Quantiles of full video used for scale normalization
+        (default: None)
+
+    min_max: Optional[Tuple[float, float][
+        Minimum and maximum values used for scale normalization
+        (default: None)
+
+    time_chunk_size: int
+        Number of time steps to process at once.
+        (default: 100)
+
+    Returns
+    -------
+    data: np.ndarray
+        Video, cropped to the specified field of view and scaled
+        to np.uint8 (i.e. dynamic range is [0, 255])
+
+    Notes
+    -----
+    One and only one of quantiles, min_max must be specified. If
+    both or neither are specified, a RuntimeError will be raised.
+    """
+
+    if quantiles is None and min_max is None:
+        raise RuntimeError("must specify either quantiles or min_max "
+                           "in _read_and_scale_by_chunk; both are None")
+    if quantiles is not None and min_max is not None:
+        raise RuntimeError("cannot specify both quantiles and min_max "
+                           "in _read_and_scale_by_chunks")
+
+    with h5py.File(full_video_path, 'r') as in_file:
+        dataset = in_file['data']
+        rowmin = origin[0]
+        rowmax = min(dataset.shape[1], origin[0]+frame_shape[0])
+        colmin = origin[1]
+        colmax = min(dataset.shape[2], origin[1]+frame_shape[1])
+
+        if quantiles is not None:
+            min_max = np.quantile(dataset[()], quantiles)
```

Comment:
The thing I am trying to avoid is having two copies of the whole dataset in memory at once.

A few lines down, I need to create a uint8 version of the dataset, and populate it from the normalized chunks of the int16 dataset. Reading in the entire dataset once here isn't necessarily a problem (until we start handling those 100 GB movies I keep fretting about). The problem comes when I construct the empty array of np.uint8s. If this dataset is still in memory, then I will have 2 (really 1.5) complete copies of the movie in memory at once, which can be prohibitive, especially once we start manipulated them in `normalize_array`. Ideally, the user will specify `min_max` rather than `quantiles` in this method, but I don't want to hamstring them this way.

it's not I/O efficient to read the dataset in twice, but it is memory efficient (if only just).

Reply:
I'll add a comment explaining my thinking